### PR TITLE
crypto/openssl_*: Standardize implementaton and add support keyUsage, extenededKeyUsage

### PR DESCRIPTION
**Pull request by cpointner**
##### SUMMARY

The OpenSSLObject class has been merged[1]. This commit makes the
openssl_csr module rely on this class and standardize the way openssl
module should be written.
This also adds checks for subject alt names as well as support for keyUsage
and extendedKeyUsage certificate extensions.

This is based on the work of @psykotox and obsoletes #25215.

[1] #26945

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

  - openssl_csr

##### ANSIBLE VERSION
```
devel
```

##### ADDITIONAL INFORMATION
```
N/A
```

### CODE REVIEW

File: lib/ansible/modules/crypto/openssl_csr.py
Comment (Spredzy) [inline lib/ansible/modules/crypto/openssl_csr.py:129]: Although to mention that it will `verify()` the privatekey it has been generated with.

File: lib/ansible/modules/crypto/openssl_csr.py
Comment (Spredzy) [inline lib/ansible/modules/crypto/openssl_csr.py]: This isn't needed anymore as it is part of the parent class value. **Note**: Default has been set to `False`, so the logique about self.changed needs to be updated in the rest of the code.

File: lib/ansible/modules/crypto/openssl_csr.py
Comment (Spredzy) [inline lib/ansible/modules/crypto/openssl_csr.py:315]: Not sure I understand the `self.privatekey`` here.

I'd say it belongs circa l. 367 within the `def _check_signature()` method as this is the only place where it is need.

File: lib/ansible/modules/crypto/openssl_csr.py
Comment (Spredzy) [inline lib/ansible/modules/crypto/openssl_csr.py]: nit: can you:

 * rename `csr` to `filename` to matches privatekey and publickey module. So as the user, one would always know what is the return parameter s/he needs to look at.
 * add 'privatekey': self.privatekey_path, so a user would know which privatekey has been used to generate that csr

**Review by Spredzy — CHANGES_REQUESTED** [review]
Thanks @cpointner for taking up on this work. First review is done, minor changes required. I'll just dig deeper on the KeyUsage and extendedKeyUsage on my side, just to ensure I get everything you're doing on the check.

In general, LGTM on this review after comments are addressed

File: lib/ansible/modules/crypto/openssl_csr.py
Comment (cpointner) [reply lib/ansible/modules/crypto/openssl_csr.py:315]: No it is also needed when signing the certificate request (see L289) and i think it is ugly to open the private key twice. Opening it in __init__ is of course a very bad idea so this is actually the only place where i can be sure that the key gets opened before an existing CSR gets checked or a new one is signed.

File: lib/ansible/modules/crypto/openssl_csr.py
Comment (Spredzy) [inline lib/ansible/modules/crypto/openssl_csr.py:340]: This check doesn't work as-is and raises issues when running the following playbook two times in a row:

```
---
- hosts: localhost
  tasks:
    - openssl_privatekey:
        path: /tmp/private.key

    - openssl_csr:
        path: /tmp/csr.csr
        privatekey_path: /tmp/private.key                                                                                                                                                      
        commonName: www.ansible.com
```
This is due to the fact thatthe current code relies on `expected` being an array when it could be actually None, hence raising:

```
TypeError: 'NoneType' object is not iterable
```

Here is an alternative implentation that did what was expected, feel free to modify adapt/modify/get ideas from it:

```
usages_ext = [str(ext) for ext in extensions if ext.get_short_name() == extName]

if (not usages_ext and expected) or (usages_ext and not expected):
    return False
elif not usages_ext and not expected:
    return True
else:
    current = [usage.strip() for usage in usages_ext[0].split(',')]
    expected = [long[usage] for usage in expected]
    return current == expected
```

File: lib/ansible/modules/crypto/openssl_csr.py
Comment (Spredzy) [reply lib/ansible/modules/crypto/openssl_csr.py:315]: Fair engough. Then, if I haven't overlooked anything, there is no need to pass `privatekey` to `_check_signature()` as you can call `csr.verify(self.privatekey)` directly.

**Review by Spredzy — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/crypto/openssl_csr.py
Comment (cpointner) [reply lib/ansible/modules/crypto/openssl_csr.py:315]: No i can't because `_check_signature` as all the other _check functions is no member function of the class

File: lib/ansible/modules/crypto/openssl_csr.py
Comment (cpointner) [reply lib/ansible/modules/crypto/openssl_csr.py:340]: I more or less used your proposed check function. It is indeed much nicer than my version.
I only had to slightly change the list comprehension for `expected`. The reason for this is that the user supplied value in `expected` might already be the long version of the usage name. Openssl accepts both the long and the short version when creating the CSR but will always returns the long version. For this reason i need to look up the long name for usages where we got the short name from the user.

**Comment by cpointner** [comment]
Closing & reopening to trigger CI

File: lib/ansible/modules/crypto/openssl_csr.py
Comment (cpointner) [reply lib/ansible/modules/crypto/openssl_csr.py:315]: ...which is wrong of course, because the _check functions are closures. Stupid me... Sorry for the noise!

**Comment by cpointner** [comment]
@Spredzy thanks for your review. I think i have addressed all of your comments. If you prefer i can squash my changes into one or two commits to make it easier to review the changes.

File: lib/ansible/modules/crypto/openssl_csr.py
Comment (Spredzy) [inline lib/ansible/modules/crypto/openssl_csr.py]: Not your fault, but can you remove the aliases[] please, since its the same value

File: lib/ansible/modules/crypto/openssl_csr.py
Comment (Spredzy) [inline lib/ansible/modules/crypto/openssl_csr.py:378]: Can you updated the RETURN documentation to:

  * s/csr/filename
  * add `keyUsage` and `extendedUsage`

**Review by Spredzy — CHANGES_REQUESTED** [review]
Two minor things and then shipit!

For the commit squash do not worry about it, I think that the bot or the merge strategy will eventually squash them all in one.

**Comment by cpointner** [comment]
bot_status

**Comment by cpointner** [comment]
ready_for_review

**Comment by Spredzy** [comment]
shipit

**Comment by Spredzy** [comment]
bot_status

**Comment by cpointner** [comment]
@Spredzy: i'm a little confused... the bot says it's waiting for me but i don't know what needs to be done here. Do you know whats missing?

**Review by Spredzy — APPROVED** [review]

**Comment by Spredzy** [comment]
bot_status

**Comment by Spredzy** [comment]
@cpointner The waiting_on: maintainer should be solved once https://github.com/ansible/ansible/pull/27691 is merged.

This is because `modules_utils/crypto.py` is modified and I am not listed as the maintainer, hence the message. Will redo a bot_status once merged

**Comment by Spredzy** [comment]
bot_status

**Review by gundalow — APPROVED** [review]

**Comment by gundalow** [comment]
bot_status

**Comment by Spredzy** [comment]
Thanks @cpointner for this work. Much appreciated :)

**Pull request merged**
